src/debug.c

```diff
@@ -0,0 +1,603 @@
+	return STA_NOINIT;
+}
+
+DSTATUS disk_status(BYTE pdrv)
+{
+	if (fat_disks[pdrv].disk_status)
+		return fat_disks[pdrv].disk_status();
+	return STA_NOINIT;
+}
+
+DRESULT disk_read(BYTE pdrv, BYTE* buff, LBA_t sector, UINT count)
+{
+	if (fat_disks[pdrv].disk_read)
+		return fat_disks[pdrv].disk_read(buff, sector, count);
+	return RES_PARERR;
+}
+
+DRESULT disk_write(BYTE pdrv, const BYTE* buff, LBA_t sector, UINT count)
+{
+	if (fat_disks[pdrv].disk_write)
+		return fat_disks[pdrv].disk_write(buff, sector, count);
+	return RES_PARERR;
+}
+
+DRESULT disk_ioctl(BYTE pdrv, BYTE cmd, void* buff)
+{
+	if (fat_disks[pdrv].disk_ioctl)
+		return fat_disks[pdrv].disk_ioctl(cmd, buff);
+	return RES_PARERR;
+}
+
+/*********************************************************************
+ * FAT backend: 64drive
+ *********************************************************************/
+
+#define D64_CIBASE_ADDRESS   0xB8000000
+#define D64_BUFFER           0x00000000
+#define D64_REGISTER_STATUS  0x00000200
+#define D64_REGISTER_COMMAND 0x00000208
+#define D64_REGISTER_LBA     0x00000210
+#define D64_REGISTER_LENGTH  0x00000218
+#define D64_REGISTER_RESULT  0x00000220
+
+#define D64_CI_IDLE  0x00
+#define D64_CI_BUSY  0x10
+#define D64_CI_WRITE 0x20
+
+#define D64_COMMAND_SD_READ  0x01
+#define D64_COMMAND_SD_WRITE 0x10
+#define D64_COMMAND_SD_RESET 0x1F
+#define D64_COMMAND_ABORT    0xFF
+
+// Utility functions for 64drive communication, defined in usb.c
+extern int8_t usb_64drive_wait(void);
+extern void usb_64drive_setwritable(int8_t enable);
+
+static DRESULT fat_disk_read_64drive(BYTE* buff, LBA_t sector, UINT count)
+{
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	for (int i=0;i<count;i++)
+	{	
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_LBA, sector+i);
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_SD_READ);
+		if (usb_64drive_wait() != 0)
+		{
+			// Operation is taking too long. Probably SD was not inserted.
+			// Send a COMMAND_ABORT and SD_RESET, and return I/O error.
+			// Note that because of a 64drive firmware bug, this is not
+			// sufficient to unblock the 64drive. The USB channel will stay
+			// unresponsive. We don't currently have a workaround for this.
+			io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_ABORT);
+			usb_64drive_wait();
+			io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_SD_RESET);
+			usb_64drive_wait();
+			return FR_DISK_ERR;
+		}
+
+		if (((uint32_t)buff & 7) == 0) 
+		{
+			data_cache_hit_writeback_invalidate(buff, 512);
+			dma_read(buff, D64_CIBASE_ADDRESS + D64_BUFFER, 512);
+		}
+		else
+		{
+			typedef uint32_t u_uint32_t __attribute__((aligned(1)));
+
+			uint32_t* src = (uint32_t*)(D64_CIBASE_ADDRESS + D64_BUFFER);
+			u_uint32_t* dst = (u_uint32_t*)buff;
+			for (int i = 0; i < 512/16; i++) 
```

Comment:
The ROM address space on N64 can only be accessed with aligned 32-bit reads/writes (or via DMA). So when DMA cannot be used because the target buffer is not aligned, then we need to do a manual copy. We cannot use `memcpy` because `memcpy` does not guarantee that all writes will be 32 bits, it expects to be able to access the memory with reads/writes of any width. Notice that the loop is also using an unaligned uin32_t type for the buffer in RDRAM so that the compiler will generate `LWL`/`SWL` for that as expected.

This whole loop is a fallback in case `fat_disk_read_64drive` is called with an unaligned buffer, otherwise DMA is used (which is much faster). I don't know if/when that happens, it is part of fatfs internals. Most of the times I have seen calls to read the whole `win` buffer in the `FATFS` structure, so it looks like fatfs tries to read full sectors as an internal optimization. `win` should be aligned -- now that I read this code again, I don't see an `attribute((aligned(8)))` added to it, maybe I've lost that as part of the PR. I will readd it.